### ISSUE TITLE
[css-text-decor-5] Allow percentages in `text-decoration-trim`

### ISSUE THREAD

**Issue by SebastianZ** [w3c/csswg-drafts#8403]
There was already some discussion in #4557 around allowing to use percentage values in `text-decoration-trim`.

One relatively common use case is to animate an underline from no width to its normal size when hovering a link. I saw this on several sites, lately, though I didn't save the URLs, unfortunately. Nontheless, here are some blogs and fiddles showing this effect:

https://speckyboy.com/underline-text-effects-css/
https://dev.to/afif/100-underline-overlay-animation-the-ultimate-css-collection-4p40
https://css-irl.info/animating-underlines/
https://www.rustcodeweb.com/2023/10/create-animated-underline-links-on-hover-using-html-css-only.html
https://www.dannyguo.com/blog/animated-multiline-link-underlines-with-css
https://www.tutorialkart.com/css/how-to-display-animated-underline-on-hover-for-a-link-using-css/
https://codepen.io/jstn/pen/mdoOZJ

Another one is to partially underline a headline based on its width.

The main question regarding this feature was to what the percentage values relate to.
@fantasai initially defined them to be [relative to the width of the containing block for block containers and relative to the total length of the untrimmed line decoration for inline elements](https://github.com/w3c/csswg-drafts/issues/4557#issuecomment-1117735704).
As decorations don't extend to the edges of a block container, I argued that decoration and text lengths would be disconnected with that definition. My suggestion was to [let them refer to the width of each fragment of the decoration box or the total length of the untrimmed line decoration of the in-flow inline-level children](https://github.com/w3c/csswg-drafts/issues/4557#issuecomment-1117950076).
@jfkthame and @litherum then mentioned that referring to the fragment widths might also not be the best approach and mentioned issues regarding [short fragments vs. long fragments](https://github.com/w3c/csswg-drafts/issues/4557#issuecomment-1176014387) and [bidi runs](https://github.com/w3c/csswg-drafts/issues/4557#issuecomment-1176744075).
@dbaron then [noted](https://drafts.csswg.org/css-break/#break-decoration) that this might be controlled via [`box-decoration-break`](https://drafts.csswg.org/css-break/#break-decoration).

Sebastian

**Comment by SebastianZ** [comment]
Oh, of course, one well-known site that has an underline animation is https://2022.stateofcss.com/.

Sebastian

**Comment by SebastianZ** [comment]
The example page I initially posted doesn't use animated underlines anymore, so I've replaced the link by some blog posts showing the effect.

To make progress, I'd say we should **let percentages refer to the inline size of the [decorating box](https://drafts.csswg.org/css-text-decor-4/#decorating-box). And if the decoration is broken into several box fragments and the decoration is meant to be cloned, the percentages refer to the inline size of each box fragment.**

Regarding breaking the decoration, we might either reuse `box-decoration-break` or introduce a new property `text-decoration-break` that only affects text decorations and would take the same values as `box-decoration-break`. (See #12947)

~~I think the definition of `text-decoration-trim` is currently still unclear regarding bidi runs. And that's independent of allowing percentages, i.e. we should discuss that separately from this issue.~~

AFAIU, with the current definition, `text-decoration-trim` follows the writing mode, but bidi runs don't affect the trimming.

Sebastian

**Comment by yisibl** [comment]
Underline animations are definitely a very common requirement!

@jfkthame I see Firefox has already [implemented](https://bugzilla.mozilla.org/show_bug.cgi?id=1988793) the `text-decoration-trim` property. Could we consider supporting percentage values before shipping?

**Comment by Loirooriol** [comment]
@yisibl It's not clear at all how percentages resolve. Why should this block shipping? It can always be added later.

**Comment by yisibl** [comment]
@Loirooriol I just hope it would be better if we could resolve this issue before Firefox shipping.

Of course, as things stand, we still have many details to discuss.

**Comment by SebastianZ** [comment]
I agree that this may not block shipping, though it would still be nice if we could resolve on a proper behavior.

So, I imagine [my proposal](https://github.com/w3c/csswg-drafts/issues/8403#issuecomment-3396895682) to render like in the following examples:

One-value case:

```
data:text/html,<p style="width:10em;font-size:3em;">This is a text including
<span style="text-decoration:0.1em underline;text-underline-offset:0.25em;
text-decoration-trim:10%;">some underlined part that wraps</span>
into two lines.</p>
```

would render like this:

<img width="240" height="112" alt="Rendering of text-decoration-trim with a single value of 10%" src="https://github.com/user-attachments/assets/8baa857e-1ccb-45fe-8904-1c3fb7c19213" />

Two-value case:

```
data:text/html,<p style="width:10em;font-size:3em;">This is a text including
<span style="text-decoration:0.1em underline;text-underline-offset:0.25em;
text-decoration-trim:0 20%;">some underlined part that wraps</span>
into two lines.</p>
```

<img width="240" height="112" alt="Rendering of text-decoration-trim with two values, the first one 0, the second one 20%" src="https://github.com/user-attachments/assets/5268abee-1098-41b9-8820-235ca145a6cd" />

`box-decoration-break` set to `clone`:

```
data:text/html,<p style="width:10em;font-size:3em;">This is a text including
<span style="text-decoration:0.1em underline;text-underline-offset:0.25em;
text-decoration-trim:5%;box-decoration-break:clone;">some underlined part that wraps</span>
into two lines.</p>
```

<img width="240" height="112" alt="Rendering of text-decoration-trim with a single value of 10% and box-decoration-break set to clone" src="https://github.com/user-attachments/assets/789fe28d-a04b-4ce2-927d-7dc24e9c89d7" />

So given that, maybe we can get a resolution. So adding this to the agenda.

-----

What has to be discussed before shipping are #8402 and #12947.

Sebastian

**Comment by jfkthame** [comment]
I don't see any clarity about how percentages would resolve. As such, I think it would be premature to ship support for percentages. But it can be added later as an enhancement once we have a clear idea of what it means.

**Comment by SebastianZ** [comment]
I've updated [my proposal](https://github.com/w3c/csswg-drafts/issues/8403#issuecomment-3396895682) to make it clearer how percentages should behave in regard of `box-decoration-break` being set to `slice` vs. `clone` and fixed my example related to it in [my previous comment](https://github.com/w3c/csswg-drafts/issues/8403#issuecomment-3448944643).

@jfkthame @Loirooriol Please let me know whether it's still unclear to you!

Sebastian

**Comment by SebastianZ** [comment]
As Mozilla is actively working on an implementation of `text-decoration-trim` and current work happens on level 4, I've switched the labels.

Sebastian

**Comment by LeaVerou** [comment]
The more I think about this, the more I become convinced that the solution to _both_ this and #8402 is `text-decoration-length: <length-percentage>` with an initial value of `100%`.

You want outset?

```css
text-decoration-length: calc(100% + .2em);
```

You want inset?

```css
text-decoration-length: calc(100% - .2em);
```

Yes, it’s 11 characters longer than `text-decoration-inset: .2em` but IMO the clarity and flexibility is worth it.

You want to animate from no underline? You can!

```css
@keyframes underline-grow {
	from { text-decoration-length: 0 }
}

.foo {
	animation: underline-grow 1s both;
}
```

You want to animate with a constant speed for each line (when `text-decoration-break` is `clone`)? You _also_ can!

```css
@property --text-decoration-max-length {
	syntax: "<length-percentage>";
	inherits: false;
	initial-value: 0px;
}

@keyframes underline-grow {
	to { --text-decoration-max-length: 100cqw }
}

.foo {
	animation: underline-grow 1s both;
	text-decoration-length: min(100%, var(--text-decoration-max-length);
}
```

We had similar discussions when we were defining relative color syntax, and in the end we went with the flexibility and clarity of  `calc()` for similar reasons (and I think it was an excellent decision).

**Comment by jfkthame** [comment]
In this model, how do you inset (or outset) one end of the line independently of the other?

**Comment by jfkthame** [comment]
> You want to animate with a constant speed for each line (when `text-decoration-break` is `clone`)? You _also_ can!
> 
> @keyframes underline-grow {
> 	from { text-decoration-length: 0 }
> 	to { text-decoration-length: min(100%, 100cqw) }
> }
> 
> .foo {
> 	animation: underline-grow 1s both;
> }

It may just be my lack of understanding, but I don't see how this gives constant-speed animation if a decorated element is broken into two fragments with quite different widths. The `100%` will resolve to different lengths for each fragment, and so the animation will grow at different speeds such that they each reach 100% after 1s, no? Or are there some other rules for how this works?

**Comment by LeaVerou** [comment]
> In this model, how do you inset (or outset) one end of the line independently of the other?

I don't think you can (except perhaps through another property to offset). Are there use cases that need this? Do other models allow this?

> > You want to animate with a constant speed for each line (when `text-decoration-break` is `clone`)? You _also_ can!
> > [@Keyframes](https://github.com/Keyframes) underline-grow {
> > from { text-decoration-length: 0 }
> > to { text-decoration-length: min(100%, 100cqw) }
> > }
> > .foo {
> > animation: underline-grow 1s both;
> > }
> 
> It may just be my lack of understanding, but I don't see how this gives constant-speed animation if a decorated element is broken into two fragments with quite different widths. The `100%` will resolve to different lengths for each fragment, and so the animation will grow at different speeds such that they each reach 100% after 1s, no? Or are there some other rules for how this works?

Good catch! Just updated the example.

**Comment by dbaron** [comment]
> > In this model, how do you inset (or outset) one end of the line independently of the other?
> 
> I don't think you can (except perhaps through another property to offset). Are there use cases that need this? Do other models allow this?

I could imagine wanting to adjust the offset asymmetrically as a sort of "italic correction" for text that's both italic and underlined, particularly in a case where the underline offset is also adjusted.  That said, it's just something I thought of just now...

**Comment by jfkthame** [comment]
> > In this model, how do you inset (or outset) one end of the line independently of the other?
> 
> I don't think you can (except perhaps through another property to offset). Are there use cases that need this? Do other models allow this?

The current spec for `text-decoration-inset` (formerly `-trim`) accepts two `<length>s`, to adjust the start and end of the decoration line. So (assuming percentage support) you could, for example, animate the inset from `0 100%` to `0 0` to make the line grow from the start of the element to its end. Or you can set the insets to `1em -1em` to shift the line 1em forward without changing its length, etc. I can imagine a designer wanting a line to oscillate gently from side to side (e.g. as a `:hover` effect), which is easy with the `text-decoration-inset` model.

**Comment by jfkthame** [comment]
> > > You want to animate with a constant speed for each line (when text-decoration-break is clone)? You also can! ...

> Good catch! Just updated the example.

I still don't see how that works. The new version:
```
@property --text-decoration-max-length {
	syntax: "<length-percentage>";
	inherits: false;
	initial-value: 0px;
}

@keyframes underline-grow {
	to { --text-decoration-max-length: 100cqw }
}

.foo {
	animation: underline-grow 1s both;
	text-decoration-length: max(100%, var(--text-decoration-max-length);
}
```
looks to me like it would grow the underline to the full width of `100cqw`, regardless of the length of the fragment being underlined, which is not the intended result.

**Comment by Loirooriol** [comment]
If we have a property to control the size, then we need another property to control the position/aligment. Because several people will want it centered, several people will want it aligned to the start, several to the end, and some people will want other things.

It could be something like `background-position`, where 50% means center, 0% means left/start, 100% means right/end. And also combinations of `<length-percentage>` with side references, like `start 10px` or `end 10px`.

I agree with @jfkthame that this approach doesn't seem to allow constant-speed animations.

**Comment by LeaVerou** [comment]
@jfkthame I just corrected `max()` to `min()`. Does it make sense now?

> The current spec for `text-decoration-inset` (formerly `-trim`) accepts two `<length>s`, to adjust the start and end of the decoration line. So (assuming percentage support) you could, for example, animate the inset from `0 100%` to `0 0` to make the line grow from the start of the element to its end. Or you can set the insets to `1em -1em` to shift the line 1em forward without changing its length, etc. I can imagine a designer wanting a line to oscillate gently from side to side (e.g. as a `:hover` effect), which is easy with the `text-decoration-inset` model.

Ah, that is quite interesting! Though if the intent is to shift, there should be a way to express that, expressing it as two lengths is a workaround as it requires authors to transform their intent (to be fair, so does my constant speed example, but _all_ constant speed animations in CSS are hacky, since CSS animations are specified through overall time, not speed).

**Comment by Loirooriol** [comment]
> Does it make sense now?

No. How is that supposed to work? Unless there is overflow, 100% won't be greater than 100cqw, so the minimum will just be 100%. I don't see how adding this min() with a registered property would do anything.

**Comment by LeaVerou** [comment]
> > Does it make sense now?
> 
> No. How is that supposed to work? Unless there is overflow, 100% won't be greater than 100cqw, so the minimum will just be 100%. I don't see how adding this min() with a registered property would do anything.

- Percentages resolve based on the decorating box (see https://github.com/w3c/csswg-drafts/issues/8403#issuecomment-3396895682)
- Animating to `100cqw` means every line animates at the same speed because it's animating between the same start and end value
- Registering the property makes it interpolatable
- `min(100%, var(--text-decoration-max-length))` ensures that we never get more than the max line length. Shorter lines would just stop earlier.

It's basically the same way people used to animate to `height: auto` (via `max-height`) before we got `calc-size()`.

**Comment by Loirooriol** [comment]
Ah, I see it now, I hadn't looked at it correctly. I guess that would work indeed.

**Comment by jfkthame** [comment]
OK, thanks - yeah, now that makes sense.

FWIW, I think essentially the same thing could be done with the `text-decoration-inset` version (once percentages are supported). All the same information is available as inputs to the animation. It'd just be used in a different version of the calculation, but whatever can be calculated as a length for the line can also be calculated as an amount to inset from the original length.

I think, though, that for the common use cases, the currently-spec'd `-inset` (formerly `-trim`) property is the more author-friendly approach, applying the specified adjustment(s) to the ends of the standard decoration line. Requiring an author to resort to `calc()` just to extend each end of the line by a fixed amount feels overly cumbersome to me.

**Comment by SebastianZ** [comment]
I believe we are already past the discussion about `text-decoration-length` vs. `text-decoration-inset`. Both have their pros and cons, we previously resolved on the latter and Firefox already shipped it.

To be clear, @LeaVerou's examples (with a centered underline) look like this using `text-decoration-inset`:

Animating from no underline:

```css
@keyframes --underline-grow {
  from { text-decoration-inset: 50% }
}

.foo {
  text-decoration: underline;
  animation: --underline-grow 1s both;
}
```

Animating with constant speed:

```css
@keyframes --underline-grow {
  from { text-decoration-inset: 50cqw }
}

.foo {
  text-decoration: underline;
  animation: --underline-grow 1s both;
  box-decoration-break: clone;
}
```

Plus growing at the same start time:

```css
@property --text-decoration-max-length {
  syntax: "<length-percentage>";
  inherits: false;
  initial-value: 0px;
}

@keyframes --underline-grow {
  to { --text-decoration-max-length: 100cqw }
}

.foo {
  animation: --underline-grow 1s both;
  text-decoration-inset: max(0, calc(100% - var(--text-decoration-max-length) / 2));
}
```

Therefore, both use cases of constant time and constant speed of growing are covered by that. And it seems we already agreed on percentages referring to the total width of decoration box fragments for `box-decoration-break: slice` and the width of each individual decoration box fragment for `box-decoration-break: clone`, right?

So let's make a second attempt to resolve on that.

Sebastian

**Comment by svgeesus** [comment]
Closed by https://github.com/w3c/csswg-drafts/pull/14269

**Comment by SebastianZ** [comment]
#14269 actually doesn't cover what we resolved on. So opening again.

Sebastian

**Comment by SebastianZ** [comment]
Closing again, since the new PR #14282 got merged.

Sebastian
